### PR TITLE
Add configurability for max header field size in HPACK

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/aspnet/KestrelHttpServer/issues/2872

### CODE REVIEW

File: src/Kestrel.Core/Http2Limits.cs [hunk 1]
```diff
@@ -74,5 +75,25 @@ public int MaxFrameSize
...
             }
         }
+
+        /// <summary>
+        /// Indicates the size of the maximum allowed size of a header field sequence. This limit applies to both name and value sequences in their compressed and uncompressed representations.
+        /// <para>
+        /// Value must be greater than 0, defaults to 4096
+        /// </para>
+        /// </summary>
+        public int MaxHeaderFieldSize
```
Comment (JunTaoLuo) [inline src/Kestrel.Core/Http2Limits.cs]: Technically `header field` refers to the name-value pair according to the RFC. But they didn't have a specific term for the name or value portion of the header field and I'd rather keep the name as simple as possible.

File: test/Kestrel.Core.Tests/HPackDecoderTests.cs [hunk 2]
```diff
@@ -442,10 +443,28 @@ public void DecodesStringLength_GreaterThanLimit_Error()
...
         }
 
+        [Fact]
+        public void DecodesStringLength_LimitConfigurable()
+        {
+            var decoder = new HPackDecoder(DynamicTableInitialMaxSize, 4097);
+            var string4097 = new string('a', 4097);
+
+            var encoded = _literalHeaderFieldWithoutIndexingNewName
+                .Concat(new byte[] { 0x7f, 0x82, 0x1f }) // 4097 encoded with 7-bit prefix, no Huffman encoding
```
Comment (halter73) [inline test/Kestrel.Core.Tests/HPackDecoderTests.cs]: The `DecodesStringLength_GreaterThanLimit_Error` test is different.

```
var encoded = _literalHeaderFieldWithoutIndexingNewName
    .Concat(new byte[] { 0xff, 0x82, 0x1f }) // 4097 encoded with 7-bit prefix
```

Shouldn't these match? I'm guessing the old test is wrong since this test actually verifies the dictionary.

File: test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs [hunk 3]
```diff
@@ -331,12 +331,27 @@ public void Http2HeaderTableSizeDefault()
         }
 
         [Theory]
-        [InlineData(4097)]
+        [InlineData(int.MinValue)]
         [InlineData(-1)]
         public void Http2HeaderTableSizeInvalid(int value)
         {
-            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => new KestrelServerLimits().Http2.MaxFrameSize = value);
```
Comment (halter73) [inline test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs]: I see the old test was improperly named, but we shouldn't remove coverage for MaxFrameSize if this isn't tested elsewhere. The old InlineData values should still be invalid for MaxFrameSize.

File: test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs [hunk 4]
```diff
@@ -331,12 +331,27 @@ public void Http2HeaderTableSizeDefault()
         }
 
         [Theory]
-        [InlineData(4097)]
+        [InlineData(int.MinValue)]
         [InlineData(-1)]
```
Comment (halter73) [inline test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs]: Might as well include 0 as an edge case.

File: src/Kestrel.Core/Http2Limits.cs (hunk 1 above)
Comment (halter73) [reply src/Kestrel.Core/Http2Limits.cs]: This is only for request headers, right? We should probably make that clear. We should probably name this MaxRequestHeaderFieldSize.

**Comment by JunTaoLuo** [comment]
Updated

File: src/Kestrel.Core/Http2Limits.cs [hunk 5]
```diff
@@ -14,11 +14,12 @@ public class Http2Limits
         private int _maxStreamsPerConnection = 100;
         private int _headerTableSize = (int)Http2PeerSettings.DefaultHeaderTableSize;
         private int _maxFrameSize = (int)Http2PeerSettings.DefaultMaxFrameSize;
+        private int _maxRequestHeaderFieldSize = 4096;
```
Comment (Tratcher) [inline src/Kestrel.Core/Http2Limits.cs]: It should be at least 8kb so it can handle the same request path & query values as HTTP/1

File: test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs (hunk 3 above)
Comment (halter73) [reply test/Kestrel.Core.Tests/KestrelServerLimitsTests.cs]: I see. You might as well fix the inline data for the Http2MaxFrameSizeInvalid test. `1 << 14 - 1` should be `(1 << 14) - 1`.
